core/src/main/java/com/google/errorprone/bugpatterns/AbortInOvercatch.java

```diff
@@ -0,0 +1,138 @@
+ *   // stmt 2 throws exception A
+ * } catch (Throwable t) {
+ *   // more than one unique exceptions fall into this catch -- overcatch
+ *   abort(); 
+ * }
+ * </pre>
+ *
+ * While it is a best practice to catch the precise exceptions, abort or System.exit()
+ * in exception over-catches are particularly dangerous as it might accidentally
+ * bring down the cluster on unexpected scenarios. 
+ * 
+ * Because it is difficult to check for the exceptions each statement throws in 
+ * AST, we take the liberty to simplify our implementation: we simply check if the
+ * catch block catches "Throwable" or "Exception", and then calls abort(). 
+ * 
+ * For more detail, refer to the paper:
+ * "Simple Testing Can Prevent Most Critical Failures: 
+ *  An Analysis of Production Failures in Distributed Data-intensive Systems"
+ *  Yuan et al. Proceedings of the 11th Symposium on Operating Systems Design 
+ *  and Implementation (OSDI), 2014
+ *
+ * @author yuan@eecg.utoronto.ca (Ding Yuan)
+ */
+@BugPattern(name = "AbortInOvercatch",
+    summary = "Aborting the system after catching Throwable or Exception. "
+            + "Because these exceptions are high-level exceptions, this might accidentally"
+            + " take down the systems on unexpected exceptions.",
+    explanation = "Aborting the system in Throwable or Exception is bad practice.",
+    category = JDK, maturity = MATURE, severity = ERROR)
+public class AbortInOvercatch extends BugChecker implements TryTreeMatcher {
+  @Override
+  public Description matchTry (TryTree tree, VisitorState state) {
+    List<? extends CatchTree> catchList = tree.getCatches();
+    if (catchList == null || catchList.size() == 0) {
+      // TODO: this try block does not have a catch, we should further check the 
+      // finally block!
+      return Description.NO_MATCH;
+    }
+ 
+    CatchTree lastCatch = catchList.get(tree.getCatches().size() - 1);
+    if (overcatch(lastCatch, state)) {
+      if (abortInCatch(lastCatch, state)) {
+        LineMap lineMap = state.getPath().getCompilationUnit().getLineMap();
+
+        return describeMatch(lastCatch);
+      }
+    }
+    return Description.NO_MATCH;
+  }
+
+  private boolean overcatch(CatchTree catchTree, VisitorState state) {    
+    String caughtException = catchTree.getParameter().getType().toString();
+    // System.out.println("DEBUG: Caught exception: " + caughtException);
+    if (caughtException.equals("Throwable") || caughtException.equals("Exception")) {
+      return true;
+    }
+    return false;
+  }
+  
+  private boolean abortInCatch(CatchTree catchTree, VisitorState state) {
+    List<? extends StatementTree> statements = catchTree.getBlock().getStatements();
+    if (statements.isEmpty()) {
+      return false;
+    }
+    StatementTree lastStmt = statements.get(statements.size() - 1);
+    if (lastStmt.getKind() == EXPRESSION_STATEMENT) {
+      ExpressionTree et = ((ExpressionStatementTree) lastStmt).getExpression();
+
```

Comment:
Here I think you can use our MethodMatchers library to do the same thing in a clearer way.  It would be something like:
```
private static final Matcher<ExpressionTree> ABORT_MATCHER = Matchers.anyOf(
  MethodMatchers.staticMethod.onClass("java.lang.System").named("exit"),
  MethodMatchers.anyMethod().anyClass().withNameMatching(Pattern.compile(".*(?i:abort).*")),
  MethodMatchers.anyMethod().anyClass().withNameMatching(Pattern.compile(".*(?i:shutdown).*")));
```

And then this method can just return ```ABORT_MATCHER.matches(et, state)```.